### ISSUE TITLE
[Question] Stateful learning

### ISSUE THREAD

**Issue by gloriouskilka** [bshao001/ChatLearner#24]
Hello!
Great repository!

Have you tried to save-restore encoder state from sequence to sequence to be able to have a context of dialogue?

I train my own model based on nmt, which will try to predict very long dialogues, and I split long dialogs by <author, sentence> sequences, don't know if it'll work yet.

**Comment by gloriouskilka** [comment]
Another question about exploding perplexity: I noticed that you threw away Residual layer wrappers. As from your experience, have you noticed unability to train the model with them?

My training perplexity explodes in few hundred steps with residual layers tuned on (I didn't try to use Adam optimizer). That's weird, because according to README and examples they originally used SGD too and trained models with residual layers without problems.

**Comment by bshao001** [comment]
Hello, @gloriouskilka  

Both of your questions are related to some efforts that require significant computing ability/power. 

I have not tried the approach you suggested for the first one. I did hear that type of proposal, but I have never seen or heard any luck or success of that working. My impression is that RNN and its many variations have limited memorizing capabilities regarding the word sequences: 50 or 60 might be the limitation of most such networks. There are things like hierarchical LSTM, which might be more powerful, and be able to learn context from the previous, two, or even 3 conversation pairs, but not from very early statements.

If you are not doing research work and focusing mainly creating a practical product, I would suggest you, if you want, to try to maintain the emotional feeling and/or topic states in the session (I have the session example, but have got this far). You can create the feeling and/or topic by training a separate NN classifier (a lot of work), or be simpler, manually create them using rules as I am doing; then create different set of the conversation pairs by inserting the feeling or topic tag into the question, and then creating its corresponding answer (a lot of work). At the time of inference, you insert the feeling and/or topic tag along the way of the conversation, when appropriate.

Concerning the second question, I did not try the residual layer option. However, I did try 4 layers of the network. My experience was that it did not increase the model capacity: with the same training data set, the perplexity did not decrease faster during the training when solely increasing the the layer from 2 to 4. Only increasing the number of units increases the model capacity. However, if you like that option, my feeling is that it may at least work as the idea of residual connection is to maintain the original information in the worse case. Also, gradient clipping is used in the implementation, which means the perplexity will never go up during the training. I trained my model many times, and the perplexity also smoothly decrease epoch by epoch, thanks to gradient clipping.

Hope all these help!

**Comment by gloriouskilka** [comment]
@bshao001 Thank you for the detailed response.

> Also, gradient clipping is used in the implementation

I see clipping by max_gradient_norm, but, as I understood, it explodes _slowly_ because of them. Maybe I'll need more hyperparameters tuning if I use residual wrappers. I turned them off for now, and it goes well

Now I save `encoder_state` from `dynamic_rnn`, which is initialized with `encoder_initial_state = cell.zero_state(self.batch_size, dtype)` and get `encoder_state` each training `sess.run` feeding it with the `encoder_state` saved from the previous batch run. And I prepared the input data to be sequential in terms of batching.

Hope it'll have some sort of context memory through initial encoder state connected with previous final encoder state, which hopefully will be sort of connected with even previous sequence's encoder state.
